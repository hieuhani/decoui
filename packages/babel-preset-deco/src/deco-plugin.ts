import type { ConfigAPI, NodePath, Visitor } from "@babel/core";
import type {
  Program,
  ImportDeclaration,
  ImportSpecifier,
  Identifier,
  FunctionDeclaration,
  ArrowFunctionExpression,
  FunctionExpression,
  CallExpression,
  ObjectExpression,
  ObjectProperty,
  VariableDeclaration,
  VariableDeclarator,
  BlockStatement,
  Node,
  StringLiteral,
  ArrayExpression,
} from "@babel/types";

export function decoPlugin(api: ConfigAPI) {
  return {
    name: "deco-plugin",
    visitor: {
      Program(path: NodePath<Program>) {
        let hasUseDecoImport = false;
        path.node.body.forEach((node: Node) => {
          if (
            node.type === "ImportDeclaration" &&
            (node as ImportDeclaration).source.value === "@decoui/ui" &&
            (node as ImportDeclaration).specifiers.some(
              (spec) =>
                spec.type === "ImportSpecifier" &&
                (spec as ImportSpecifier).imported &&
                ((spec as ImportSpecifier).imported as Identifier).name ===
                  "useDeco"
            )
          ) {
            hasUseDecoImport = true;
          }
        });

        if (!hasUseDecoImport) {
          const decoImport = path.node.body.find(
            (node): node is ImportDeclaration =>
              node.type === "ImportDeclaration" &&
              node.source.value === "@decoui/ui"
          );

          if (decoImport) {
            decoImport.specifiers.push({
              type: "ImportSpecifier",
              local: {
                type: "Identifier",
                name: "useDeco",
              },
              imported: {
                type: "Identifier",
                name: "useDeco",
              },
            });
          }
        }
      },
      FunctionDeclaration(path: NodePath<FunctionDeclaration>) {
        handleFunction(path);
      },
      ArrowFunctionExpression(path: NodePath<ArrowFunctionExpression>) {
        handleFunction(path);
      },
      FunctionExpression(path: NodePath<FunctionExpression>) {
        handleFunction(path);
      },
    },
  };
}

type FunctionNodePath =
  | NodePath<FunctionDeclaration>
  | NodePath<ArrowFunctionExpression>
  | NodePath<FunctionExpression>;

function handleFunction(path: FunctionNodePath) {
  // Skip if not a React component (assuming components start with capital letter)
  const functionName =
    path.node.type === "FunctionDeclaration" ? path.node.id?.name : undefined;
  if (!functionName || !/^[A-Z]/.test(functionName)) {
    return;
  }

  let hasDecoHook = false;
  const reactiveProps = new Set<string>();

  const visitor: Visitor = {
    CallExpression(callPath) {
      const callee = callPath.node.callee;
      if (callee.type === "Identifier" && callee.name === "deco") {
        const arg = callPath.node.arguments[0];
        if (arg && arg.type === "ObjectExpression") {
          arg.properties.forEach((prop) => {
            if (
              prop.type === "ObjectProperty" &&
              prop.key.type === "Identifier"
            ) {
              const key = prop.key.name;
              if (key === "dark") {
                reactiveProps.add("colorScheme");
              } else if (["sm", "md", "lg", "xl", "2xl"].includes(key)) {
                reactiveProps.add("windowWidth");
              }
            }
          });
        }
      }
    },
    VariableDeclaration(varPath) {
      if (
        varPath.node.declarations.some(
          (d) =>
            d.id.type === "Identifier" &&
            d.id.name === "deco" &&
            d.init?.type === "CallExpression" &&
            (d.init.callee as Identifier).name === "useDeco"
        )
      ) {
        hasDecoHook = true;
      }
    },
  };

  path.traverse(visitor);

  if (reactiveProps.size > 0 && !hasDecoHook) {
    const decoHookDeclaration: VariableDeclaration = {
      type: "VariableDeclaration",
      declarations: [
        {
          type: "VariableDeclarator",
          id: {
            type: "Identifier",
            name: "deco",
          },
          init: {
            type: "CallExpression",
            callee: {
              type: "Identifier",
              name: "useDeco",
            },
            arguments: [
              {
                type: "ArrayExpression",
                elements: Array.from(reactiveProps).map((prop) => ({
                  type: "StringLiteral",
                  value: prop,
                })),
              },
            ],
          },
        },
      ],
      kind: "const",
    };

    // Add the hook declaration at the beginning of the function body
    if (path.node.body.type === "BlockStatement") {
      path.node.body.body.unshift(decoHookDeclaration);
    }
  }
}
