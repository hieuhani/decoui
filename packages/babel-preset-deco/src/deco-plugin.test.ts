import { describe, expect, it } from "@jest/globals";
import { transformSync } from "@babel/core";
import { decoPlugin } from "./deco-plugin";

const transform = (code: string) => {
  return transformSync(code, {
    plugins: [decoPlugin],
    filename: "test.tsx",
    presets: [
      "@babel/preset-typescript",
      ["@babel/preset-react", { runtime: "automatic" }],
    ],
  })?.code;
};

describe("deco-plugin", () => {
  it("should add useDeco hook when deco is used with dark mode", () => {
    const code = `
      function MyComponent() {
        return (
          <View style={deco({ dark: { bg: 'gray.800' } })} />
        );
      }
    `;

    const result = transform(code);
    expect(result).toContain('const deco = useDeco(["colorScheme"])');
    expect(result).toMatchSnapshot();
  });

  it("should add useDeco hook when deco is used with dark mode and should be worked with arrow function", () => {
    const code = `
      const MyComponent = () => {
        return (
          <View style={deco({ dark: { bg: 'gray.800' } })} />
        );
      }
    `;

    const result = transform(code);
    expect(result).toContain('const deco = useDeco(["colorScheme"])');
    expect(result).toMatchSnapshot();
  });

  it("should add useDeco hook when deco is used with breakpoints", () => {
    const code = `
      function MyComponent() {
        return (
          <View style={deco({ sm: { width: 100 }, md: { width: 200 } })} />
        );
      }
    `;

    const result = transform(code);
    expect(result).toContain('const deco = useDeco(["windowWidth"])');
    expect(result).toMatchSnapshot();
  });

  it("should add useDeco hook with multiple reactive props", () => {
    const code = `
      function MyComponent() {
        return (
          <View style={deco({ 
            dark: { bg: 'gray.800' },
            sm: { width: 100 },
            md: { width: 200 }
          })} />
        );
      }
    `;

    const result = transform(code);
    expect(result).toContain(
      'const deco = useDeco(["colorScheme", "windowWidth"]);'
    );
    expect(result).toMatchSnapshot();
  });

  it("should not add useDeco hook for non-reactive props", () => {
    const code = `
      function MyComponent() {
        return (
          <View style={deco({ padding: 10, margin: 20 })} />
        );
      }
    `;

    const result = transform(code);
    expect(result).not.toContain("useDeco");
    expect(result).toMatchSnapshot();
  });

  it("should not transform non-component functions", () => {
    const code = `
      function helperFunction() {
        return deco({ dark: { bg: 'gray.800' } });
      }
    `;

    const result = transform(code);
    expect(result).not.toContain("useDeco");
    expect(result).toMatchSnapshot();
  });

  it("should not duplicate useDeco hook if already present", () => {
    const code = `
      function MyComponent() {
        const deco = useDeco(['colorScheme']);
        return (
          <View style={deco({ dark: { bg: 'gray.800' } })} />
        );
      }
    `;

    const result = transform(code);
    const matches = result?.match(/useDeco/g) || [];
    expect(matches.length).toBe(1);
    expect(result).toMatchSnapshot();
  });

  it("should handle multiple deco calls in the same component", () => {
    const code = `
      function MyComponent() {
        return (
          <>
            <View style={deco({ dark: { bg: 'gray.800' } })} />
            <Text style={deco({ sm: { fontSize: 14 }, md: { fontSize: 16 } })} />
          </>
        );
      }
    `;

    const result = transform(code);
    expect(result).toContain(
      'const deco = useDeco(["colorScheme", "windowWidth"]);'
    );
    expect(result).toMatchSnapshot();
  });
});
