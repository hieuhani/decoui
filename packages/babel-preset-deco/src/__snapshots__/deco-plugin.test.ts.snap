// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`deco-plugin should add useDeco hook when deco is used with breakpoints 1`] = `
"import { jsx as _jsx } from "react/jsx-runtime";
function MyComponent() {
  const deco = useDeco(["windowWidth"]);
  return /*#__PURE__*/_jsx(View, {
    style: deco({
      sm: {
        width: 100
      },
      md: {
        width: 200
      }
    })
  });
}"
`;

exports[`deco-plugin should add useDeco hook when deco is used with dark mode 1`] = `
"import { jsx as _jsx } from "react/jsx-runtime";
function MyComponent() {
  const deco = useDeco(["colorScheme"]);
  return /*#__PURE__*/_jsx(View, {
    style: deco({
      dark: {
        bg: 'gray.800'
      }
    })
  });
}"
`;

exports[`deco-plugin should add useDeco hook when deco is used with dark mode and should be worked with arrow function 1`] = `
"import { jsx as _jsx } from "react/jsx-runtime";
const MyComponent = () => {
  const deco = useDeco(["colorScheme"]);
  return /*#__PURE__*/_jsx(View, {
    style: deco({
      dark: {
        bg: 'gray.800'
      }
    })
  });
};"
`;

exports[`deco-plugin should add useDeco hook with multiple reactive props 1`] = `
"import { jsx as _jsx } from "react/jsx-runtime";
function MyComponent() {
  const deco = useDeco(["colorScheme", "windowWidth"]);
  return /*#__PURE__*/_jsx(View, {
    style: deco({
      dark: {
        bg: 'gray.800'
      },
      sm: {
        width: 100
      },
      md: {
        width: 200
      }
    })
  });
}"
`;

exports[`deco-plugin should handle multiple deco calls in the same component 1`] = `
"import { jsx as _jsx, Fragment as _Fragment, jsxs as _jsxs } from "react/jsx-runtime";
function MyComponent() {
  const deco = useDeco(["colorScheme", "windowWidth"]);
  return /*#__PURE__*/_jsxs(_Fragment, {
    children: [/*#__PURE__*/_jsx(View, {
      style: deco({
        dark: {
          bg: 'gray.800'
        }
      })
    }), /*#__PURE__*/_jsx(Text, {
      style: deco({
        sm: {
          fontSize: 14
        },
        md: {
          fontSize: 16
        }
      })
    })]
  });
}"
`;

exports[`deco-plugin should not add useDeco hook for non-reactive props 1`] = `
"import { jsx as _jsx } from "react/jsx-runtime";
function MyComponent() {
  return /*#__PURE__*/_jsx(View, {
    style: deco({
      padding: 10,
      margin: 20
    })
  });
}"
`;

exports[`deco-plugin should not duplicate useDeco hook if already present 1`] = `
"import { jsx as _jsx } from "react/jsx-runtime";
function MyComponent() {
  const deco = useDeco(['colorScheme']);
  return /*#__PURE__*/_jsx(View, {
    style: deco({
      dark: {
        bg: 'gray.800'
      }
    })
  });
}"
`;

exports[`deco-plugin should not transform non-component functions 1`] = `
"function helperFunction() {
  return deco({
    dark: {
      bg: 'gray.800'
    }
  });
}"
`;
